the first results we found had a lot of "outliers" in the names of the companies

as we can see in the file <code>2 findMostCitedCompanies.ipynb</code>

at the cell where there is the code:
<code>

#Sorting the dictionary by values

sorted_dict = dict(sorted(recurrences.items(), key=lambda item: item[1], reverse=True))

#Displaying the sorted dictionary

print(sorted_dict)

</code>

the most frequent ones are

<code>

{'a': 1582377, 'by': 321428, 'me': 243767, 'OpenAI': 209920, 'if': 161455, 'Google': 144378, 'A': 104661, 'no': 85504, 'Microsoft': 80406...}

</code>

other than the all right OpenAI, Google and Microsoft the other ones are, obviously, some errors in the companiesNames list

since the dataset has been retreived before twitter became X it is safe to assume that we can remove all the 1/2 letters companies from the list and see the results


In [1]:
import pandas as pd
import os
from os import path
import json
import seaborn
import matplotlib as plt
import numpy as np
import math
import requests
import pickle

In [2]:
#load the companies names
with open ("companiesNames.pickle", "rb") as f:
    companiesNames = list(pickle.load(f))
    print (len (companiesNames))

917331


In [3]:
#load the filtered dataset
dfPath = path.join(os.getcwd(), "codingFiltered_ChatGPT_Tweets_Dataset_Full.csv")
df = pd.read_csv(dfPath)

In [16]:
def remove_short_strings(lst):
    return [s for s in lst if len(str(s)) > 2]#some of the names are float or other kind of variables so we just cast them as strings

companiesNamesFiltered1 = remove_short_strings(companiesNames)
len (companiesNamesFiltered1)

916969

now we repeat the experiment in the 2 findMostCitedCompanies file and see the results

In [12]:
import string

def create_word_index_dict(list_of_strings):
    word_index_dict = {}
    for i, stringelem in enumerate(list_of_strings):
        words = stringelem.split()
        for word in words:
            # Remove punctuation from the word
            word = word.translate(str.maketrans('', '', string.punctuation))
            if word not in word_index_dict:
                word_index_dict[word] = {i: 1}
            else:
                if i not in word_index_dict[word]:
                    word_index_dict[word][i] = 1
                else:
                    word_index_dict[word][i] += 1
    return word_index_dict

In [8]:
wordsInTheDataset = create_word_index_dict(list(df["original_text"]))

In [17]:
def findSum (indictionary):
    val = 0
    for k, v in indictionary.items():
        val+=v
    return val

recurrences = {}
for companyName in companiesNamesFiltered1:
    if companyName in wordsInTheDataset:
        recurrences[companyName] = findSum (wordsInTheDataset[companyName])
        print (companyName, recurrences[companyName])

NCIS 3
Animaker 9
Sequoia 107
entomo 4
Scopely 2
Licious 1
INCITE 1
TRT 5
SonicWall 19
Ember 5
Amdocs 1
CybelAngel 3
DigitalTrust 8
Ripple 175
Freshworks 8
Chase 159
Teneo 1
Capgemini 5
Nutanix 5
Branch 35
Diageo 4
Appier 1
Inspira 1
PTC 22
Akamai 8
Flipkart 17
Dabba 2
Genesis 161
SambaNova 5
Flam 1
JMK 1
Oremus 4
Prove 98
anonymous 321
Sterling 42
Headstrong 10
Xerox 62
UST 8
Wildlife 11
FactSet 7
Activa 1
Appinventiv 4
Taskmo 1
Ideas2IT 3
Instacart 328
VMock 2
Hivemind 6
Xiaomi 39
TeamLease 2
Signify 1
Ivanti 2
Unqork 1
WCI 1
Workable 3
Grammarly 3558
Android 2174
BizBrolly 2
DataStax 5
1mg 3
Poshmark 13
Indian 1260
Ivy 88
Quality 678
Current 1689
Dream11 6
Enroute 2
AMIT 1
OnPassive 1
Tangoe 2
Codebook 4
Fanatics 11
Bayer 3
Invisalign 2
Amazon 6704
Private 341
Rotary 4
Truveta 1
Target 222
Ananya 1
ILS 2
Jumio 1
Roosh 1
Lamini 20
Mozilla 83
ByteDance 100
HSBC 23
CSE 27
mmm 17
nothing 6714
Matterport 2
ReactJS 699
LinkedIn 5050
Carrefour 16
autonomo 1
Innovate 51
Aristotle 54
Default

In [19]:
# Sorting the dictionary by values
sorted_dict = dict(sorted(recurrences.items(), key=lambda item: item[1], reverse=True))

# Displaying the sorted dictionary
print(sorted_dict)
print (len (sorted_dict))

{'Google': 131673, 'OpenAI': 85911, 'Microsoft': 42059, 'next': 27109, 'business': 24170, 'may': 21270, 'Check': 18438, 'NFT': 17225, 'New': 17221, 'around': 16561, 'Search': 13531, 'Twitter': 13486, 'between': 11886, 'PREMIUM': 11469, 'test': 10567, 'Notion': 10487, 'ability': 10129, 'One': 9039, 'add': 9028, 'easy': 8686, 'Please': 8611, 'bit': 8496, 'Tech': 8112, 'true': 7819, 'Prompt': 7439, 'Learn': 7358, 'GitHub': 7063, 'Chrome': 6963, 'Even': 6829, 'Plus': 6793, 'Some': 6737, 'nothing': 6714, 'Amazon': 6704, 'Apple': 6693, 'Bitcoin': 6584, 'imagine': 6506, 'English': 6225, 'Well': 6169, 'Today': 6140, 'Make': 5881, 'Open': 5856, 'Ticker': 5457, 'Create': 5450, 'Optimus': 5364, 'LinkedIn': 5050, 'BTC': 5033, 'AGI': 5007, 'win': 4948, 'RSI': 4863, 'Very': 4783, '101': 4740, 'Blockchain': 4707, 'Take': 4644, 'finally': 4330, 'Software': 4329, 'Wow': 4317, 'Future': 4207, 'experts': 4202, 'strong': 4195, 'yes': 4184, 'Imagine': 3798, 'Tesla': 3696, 'Spotify': 3689, 'Grammarly': 3558

the first names do, for sure make sense and the other ones are values that are neither wrong nor right

for examples:

there is a company called next but in the tweets it's probably not what they were referring to

to solve this issue i took a natural language processing approach

In [21]:
import nltk
from nltk.tag import pos_tag
from nltk.tokenize import word_tokenize

# Download NLTK resources if not already downloaded
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

# Function to check if a word is likely a proper noun (company name)
def is_proper_noun(word):
    tagged_word = pos_tag(word_tokenize(word))
    return tagged_word[0][1] == 'NNP'  # NNP indicates proper noun

# Filter out non-company names
company_names_filtered = {name: count for name, count in sorted_dict.items() if is_proper_noun(name)}

print(company_names_filtered)


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\alex1\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\alex1\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping taggers\averaged_perceptron_tagger.zip.


{'Microsoft': 42059, 'New': 17221, 'Apple': 6693, 'Human': 2482, 'York': 1941, 'John': 1575, 'Friday': 1389, 'Trust': 1338, 'Community': 1016, 'January': 1013, 'Steve': 936, 'Advanced': 863, 'IBM': 854, 'James': 803, 'San': 781, 'White': 781, 'Canada': 740, 'Africa': 729, 'Ben': 712, 'Sunday': 700, 'Alex': 693, 'Trade': 578, 'Prime': 554, 'Golden': 457, 'Minister': 413, 'Killer': 412, 'Union': 401, 'NATO': 398, 'Lee': 386, 'Smith': 382, 'Morgan': 381, 'Watson': 369, 'Uber': 365, 'Sea': 353, 'K12': 351, 'Daniel': 349, 'Richard': 344, 'Board': 344, 'Kinda': 334, 'ABC': 314, 'Harvard': 306, 'Energy': 301, 'Thomas': 295, 'Johnson': 286, 'Supreme': 278, 'Harris': 276, 'Miller': 276, 'Kong': 268, 'Alan': 244, 'BlackBerry': 238, 'OCEAN': 227, 'Mary': 226, 'Atlantic': 225, 'Knight': 222, 'Gemini': 219, 'Frank': 211, 'Charlie': 186, 'July': 174, 'Biz': 173, 'Albert': 163, 'KAL': 160, 'Joseph': 159, 'Reserve': 158, 'HUMAN': 156, 'Electric': 150, 'Golang': 146, 'Klarna': 143, 'October': 139, 'Kel

well it does quite the opposite

In [25]:
import nltk
from nltk.corpus import stopwords

# Download the stopwords corpus if you haven't already
nltk.download('stopwords')

# Get the list of stopwords
common_words = set(stopwords.words('english'))

print (len(common_words))

179


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\alex1\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


now we have 180 stopwords, let's try to filter the dictionary by it

In [26]:
# Filter out non-company names
company_names_filtered = {name: count for name, count in sorted_dict.items() if name not in common_words}
print (company_names_filtered)

{'Google': 131673, 'OpenAI': 85911, 'Microsoft': 42059, 'next': 27109, 'business': 24170, 'may': 21270, 'Check': 18438, 'NFT': 17225, 'New': 17221, 'around': 16561, 'Search': 13531, 'Twitter': 13486, 'PREMIUM': 11469, 'test': 10567, 'Notion': 10487, 'ability': 10129, 'One': 9039, 'add': 9028, 'easy': 8686, 'Please': 8611, 'bit': 8496, 'Tech': 8112, 'true': 7819, 'Prompt': 7439, 'Learn': 7358, 'GitHub': 7063, 'Chrome': 6963, 'Even': 6829, 'Plus': 6793, 'Some': 6737, 'nothing': 6714, 'Amazon': 6704, 'Apple': 6693, 'Bitcoin': 6584, 'imagine': 6506, 'English': 6225, 'Well': 6169, 'Today': 6140, 'Make': 5881, 'Open': 5856, 'Ticker': 5457, 'Create': 5450, 'Optimus': 5364, 'LinkedIn': 5050, 'BTC': 5033, 'AGI': 5007, 'win': 4948, 'RSI': 4863, 'Very': 4783, '101': 4740, 'Blockchain': 4707, 'Take': 4644, 'finally': 4330, 'Software': 4329, 'Wow': 4317, 'Future': 4207, 'experts': 4202, 'strong': 4195, 'yes': 4184, 'Imagine': 3798, 'Tesla': 3696, 'Spotify': 3689, 'Grammarly': 3558, 'Canva': 3524, '

this approach didnt work either

the next approach takes the brown corpus and orders the words in it by the occurrences

once i have this ordered dictionary i want to remove the first n (guessing around 100/1000) words that have the biggest number of occurrances

In [29]:
import nltk
from nltk.corpus import brown
from collections import Counter

# Download the Brown Corpus if not already downloaded
nltk.download('brown')

# Get the words from the Brown Corpus
words = brown.words()

# Count the occurrences of each word
word_counts = Counter(words)

# Sort the words by frequency in descending order
sorted_word_counts = sorted(word_counts.items(), key=lambda x: x[1], reverse=True)

# Create a dictionary from the sorted list
word_freq_dict = dict(sorted_word_counts)

print (word_freq_dict)

[nltk_data] Downloading package brown to
[nltk_data]     C:\Users\alex1\AppData\Roaming\nltk_data...
[nltk_data]   Package brown is already up-to-date!


{'the': 62713, ',': 58334, '.': 49346, 'of': 36080, 'and': 27915, 'to': 25732, 'a': 21881, 'in': 19536, 'that': 10237, 'is': 10011, 'was': 9777, 'for': 8841, '``': 8837, "''": 8789, 'The': 7258, 'with': 7012, 'it': 6723, 'as': 6706, 'he': 6566, 'his': 6466, 'on': 6395, 'be': 6344, ';': 5566, 'I': 5161, 'by': 5103, 'had': 5102, 'at': 4963, '?': 4693, 'not': 4423, 'are': 4333, 'from': 4207, 'or': 4118, 'this': 3966, 'have': 3892, 'an': 3542, 'which': 3540, '--': 3432, 'were': 3279, 'but': 3007, 'He': 2982, 'her': 2885, 'one': 2873, 'they': 2773, 'you': 2766, 'all': 2726, 'would': 2677, 'him': 2576, 'their': 2562, 'been': 2470, ')': 2466, '(': 2435, 'has': 2425, 'will': 2204, 'who': 2192, 'more': 2130, 'out': 2058, 'It': 2037, 'we': 1973, 'she': 1949, 'said': 1943, 'there': 1877, 'up': 1874, 'In': 1801, ':': 1795, 'than': 1788, 'them': 1786, 'into': 1782, 'no': 1781, 'its': 1780, 'about': 1766, 'so': 1755, 'when': 1746, 'can': 1738, 'only': 1646, 'other': 1627, '!': 1596, 'could': 1580, '

one of the first words that dont seem to appear at the beggining of the word_freq_dict but appears often in the sorted dictionary that we want as output is "next", i will look for the index of it in the dictionary and filter by the ones that are before it 

In [31]:
print(list(word_freq_dict.keys()).index("next"))

267


since it's 267 and we want some room we go for 300

In [32]:
def cut_dictionary(dictionary, n):
    keys = list(dictionary.keys())[:n]
    return {key: dictionary[key] for key in keys}

word_freq_dict_cutted_300 = cut_dictionary(word_freq_dict, 300)
len (word_freq_dict_cutted_300)

300

In [35]:
# Filter out non-company names
company_names_filtered_cutted_300 = {name: count for name, count in sorted_dict.items() if name not in list(word_freq_dict_cutted_300.keys())}
print (company_names_filtered_cutted_300)
print (len(company_names_filtered_cutted_300))

{'Google': 131673, 'OpenAI': 85911, 'Microsoft': 42059, 'Check': 18438, 'NFT': 17225, 'Search': 13531, 'Twitter': 13486, 'PREMIUM': 11469, 'test': 10567, 'Notion': 10487, 'ability': 10129, 'add': 9028, 'easy': 8686, 'Please': 8611, 'bit': 8496, 'Tech': 8112, 'true': 7819, 'Prompt': 7439, 'Learn': 7358, 'GitHub': 7063, 'Chrome': 6963, 'Even': 6829, 'Plus': 6793, 'Some': 6737, 'Amazon': 6704, 'Apple': 6693, 'Bitcoin': 6584, 'imagine': 6506, 'English': 6225, 'Well': 6169, 'Today': 6140, 'Make': 5881, 'Open': 5856, 'Ticker': 5457, 'Create': 5450, 'Optimus': 5364, 'LinkedIn': 5050, 'BTC': 5033, 'AGI': 5007, 'win': 4948, 'RSI': 4863, 'Very': 4783, '101': 4740, 'Blockchain': 4707, 'Take': 4644, 'finally': 4330, 'Software': 4329, 'Wow': 4317, 'Future': 4207, 'experts': 4202, 'strong': 4195, 'yes': 4184, 'Imagine': 3798, 'Tesla': 3696, 'Spotify': 3689, 'Grammarly': 3558, 'Canva': 3524, 'whatever': 3520, 'Next': 3474, 'shared': 3470, 'Edge': 3381, 'About': 3335, 'Meta': 3261, 'Build': 3087, 'Ima

the filter is not enough

so we go to the next word which is not in there

In [36]:
print(list(word_freq_dict.keys()).index("check"))

1496


In [38]:
word_freq_dict_cutted_1500 = cut_dictionary(word_freq_dict, 1500)
len (word_freq_dict_cutted_1500)

# Filter out non-company names
company_names_filtered_cutted_1500 = {name: count for name, count in sorted_dict.items() if name.lower() not in list(word_freq_dict_cutted_1500.keys())}
print (company_names_filtered_cutted_1500)
print (len(company_names_filtered_cutted_1500))

{'Google': 131673, 'OpenAI': 85911, 'Microsoft': 42059, 'NFT': 17225, 'Search': 13531, 'Twitter': 13486, 'PREMIUM': 11469, 'Notion': 10487, 'Please': 8611, 'Tech': 8112, 'Prompt': 7439, 'GitHub': 7063, 'Chrome': 6963, 'Plus': 6793, 'Amazon': 6704, 'Apple': 6693, 'Bitcoin': 6584, 'imagine': 6506, 'English': 6225, 'Ticker': 5457, 'Create': 5450, 'Optimus': 5364, 'LinkedIn': 5050, 'BTC': 5033, 'AGI': 5007, 'win': 4948, 'RSI': 4863, '101': 4740, 'Blockchain': 4707, 'Software': 4329, 'Wow': 4317, 'experts': 4202, 'yes': 4184, 'Imagine': 3798, 'Tesla': 3696, 'Spotify': 3689, 'Grammarly': 3558, 'Canva': 3524, 'shared': 3470, 'Meta': 3261, 'Netflix': 3037, 'Excel': 2926, 'Binance': 2920, 'Discover': 2866, 'Link': 2841, 'Stack': 2790, 'Disney': 2770, 'Generate': 2633, 'Assistant': 2543, 'Hello': 2455, 'Article': 2418, 'Innovation': 2403, 'connect': 2353, 'Engine': 2326, 'Model': 2277, 'Insider': 2226, 'premium': 2214, 'Scribd': 2185, 'BNB': 2179, 'YOLO': 2177, 'Reddit': 2176, 'Android': 2174, '

same results

since the results are not promising i decided to change the company names list

will continue in the 5th file